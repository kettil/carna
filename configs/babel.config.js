module.exports = {
  ignore: ['**/*.test.js', '**/*.test.ts', '**/*.test.tsx', '**/*.test-d.ts'],
  presets: ['@babel/env', '@babel/preset-typescript'],
  plugins: [
    [
      '@babel/plugin-transform-runtime', // replace helper code with runtime imports (deduplication)
      { corejs: 3 }, // import corejs polyfills exactly where they are needed
    ],
  ],
  env: {
    test: {
      // extra configuration for process.env.NODE_ENV === 'test'
      presets: ['@babel/env'], // overwrite env-config from above with transpiled module syntax
    },
  },
};
