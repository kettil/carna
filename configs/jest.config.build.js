/* eslint-disable import/no-unresolved, node/no-unpublished-require, node/no-missing-require */
module.exports = Object.assign(require('../../../jest.config.js'), {
  coverageReporters: ['text'],
});
