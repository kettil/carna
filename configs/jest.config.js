const commons = {
  bail: 10,

  testEnvironment: 'node',

  clearMocks: true,
  restoreMocks: true,
};

const createProject = (folder, color, config) => ({
  ...commons,

  displayName: { name: folder, color },
  roots: ['<rootDir>/src', `<rootDir>/tests/${folder}`],

  testMatch: ['**/*.test.{js,ts,tsx}'],
  testPathIgnorePatterns: ['/node_modules/', '/src/'],

  ...config,
});

const config = {
  ...commons,
  projects: [],

  collectCoverage: true,
  collectCoverageFrom: [
    '<rootDir>/src/**/*.{ts,tsx,js}',
    '!<rootDir>/src/**/types.ts',
    '!<rootDir>/src/index.ts',

    '<rootDir>/packages/*/src/**/*.{ts,tsx,js}',
    '!<rootDir>/packages/*/src/**/types.ts',
    '!<rootDir>/packages/*/src/index.ts',
  ],
  coveragePathIgnorePatterns: ['/__jest__/'],
  coverageReporters: ['text-summary', 'html'],
  coverageDirectory: 'coverage',
  coverageThreshold: {
    global: {
      statements: 90,
      branches: 90,
      functions: 90,
      lines: 90,
    },
  },
};

module.exports = { config, createProject };
