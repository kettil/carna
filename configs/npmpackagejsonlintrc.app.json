{
  "extends": "@kettil/npm-package-json-lint-config/app"
}
