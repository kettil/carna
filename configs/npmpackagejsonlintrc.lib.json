{
  "extends": "@kettil/npm-package-json-lint-config/lib"
}
