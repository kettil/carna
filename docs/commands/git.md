# The `git` tasks

The tasks will be automatically called by husky during commit.

## The `git commit` task

Checks the files in staged with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git msg` task

Checks the git commit message for a consistent structure.

### Options

| Options | Description                     |
| ------- | ------------------------------- |
| --edit  | Path to the COMMIT_EDITMSG file |

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |

## The `git push` task

Checks the files with type, prettier and eslint.

### Exit codes

| Exit code | Description |
| --------- | ----------- |
| 0         | Success     |
| 1         | task failed |
