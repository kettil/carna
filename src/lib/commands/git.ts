import {
  CommandModuleBuilder,
  CommandModuleDescribe,
  CommandModuleCommand,
  CommandModuleHandler,
  builderDefault,
  errorHandler,
  commonHandler,
} from '../cli/yargs';
import gitCommitTask from '../tasks/gitCommitTask';
import gitMessageTask, { GitMessageProps } from '../tasks/gitMessageTask';
import gitPushTask from '../tasks/gitPushTask';

export const command: CommandModuleCommand = 'git <hook>';
export const desc: CommandModuleDescribe = 'Handler for the git hooks';

const mode = ['msg', 'commit', 'push'] as const;
const options = { group: `${command.slice(0, Math.max(0, command.indexOf('<'))).trim()}-Options` } as const;

type Props = {
  readonly edit: GitMessageProps['edit'] | undefined;
  readonly hook: typeof mode[number];
};

export const builder: CommandModuleBuilder<Props> = builderDefault(command, (yargs) =>
  yargs.positional('hook', { choices: mode, demandOption: true }).options({
    edit: { ...options, type: 'string', desc: 'Path to the COMMIT_EDITMSG file (only at hook "msg")' },
  }),
);

export const handler: CommandModuleHandler<Props> = async (argv) => {
  try {
    const { edit, hook } = argv;

    await commonHandler(argv, false);

    switch (hook) {
      case 'commit':
        await gitCommitTask(argv);
        break;

      case 'msg':
        if (typeof edit !== 'string' || edit.trim() === '') {
          throw new Error(`Argument "edit" is required at hook "${hook}"`);
        }

        await gitMessageTask(argv, { edit });
        break;

      case 'push':
        await gitPushTask(argv);
        break;
      default:
        throw new Error(`The hook "${hook}" is unknown`);
    }
  } catch (error) {
    errorHandler(argv, error);
  }
};
