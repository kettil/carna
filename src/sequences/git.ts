import {
  CommandModuleBuilder,
  CommandModuleDescribe,
  CommandModuleCommand,
  CommandModuleHandler,
  builderDefault,
  errorHandler,
  commonHandler,
} from '../lib/cli/yargs';
import commit from './git/commit';
import msg, { Props as MessageProps } from './git/msg';
import push from './git/push';

export const command: CommandModuleCommand = 'git <hook>';
export const desc: CommandModuleDescribe = 'Handler for the git hooks';

const mode = ['msg', 'commit', 'push'] as const;
const options = { group: `${command.slice(0, Math.max(0, command.indexOf('<'))).trim()}-Options` } as const;

type Props = {
  readonly edit: MessageProps['edit'] | undefined;
  readonly hook: typeof mode[number];
};

export const builder: CommandModuleBuilder<Props> = builderDefault(command, (yargs) =>
  yargs.positional('hook', { choices: mode, demandOption: true }).options({
    edit: { ...options, type: 'string', desc: 'Path to the COMMIT_EDITMSG file (only at hook "msg")' },
  }),
);

export const handler: CommandModuleHandler<Props> = async (argv) => {
  try {
    const { edit, hook } = argv;

    await commonHandler(argv, false);

    switch (hook) {
      case 'commit':
        await commit(argv);
        break;

      case 'msg':
        if (typeof edit !== 'string' || edit.trim() === '') {
          throw new Error(`Argument "edit" is required at hook "${hook}"`);
        }

        await msg(argv, { edit });
        break;

      case 'push':
        await push(argv);
        break;
      default:
        throw new Error(`The hook "${hook}" is unknown`);
    }
  } catch (error) {
    errorHandler(argv, error);
  }
};
