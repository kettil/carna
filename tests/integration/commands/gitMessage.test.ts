import * as fs from 'fs';
import { handler } from '../../../src/lib/commands/git';
import { getArgv, getReadFileWithHooksFiles } from '../../shared/configs';

describe('command git with git "msg" hook', () => {
  test('it should work', async () => {
    const result = await handler(getArgv({ hook: 'msg', edit: 'path/to/commit.message' }));

    expect(result).toBeUndefined();
  });

  test('it should work with hooks', async () => {
    (fs as any).setMockReadFileFiles(getReadFileWithHooksFiles());

    const result = await handler(getArgv({ hook: 'msg', edit: 'path/to/commit.message' }));

    expect(result).toBeUndefined();
  });

  test('it should throw an error why "edit" is empty', async () => {
    const promise = handler(getArgv({ hook: 'msg', edit: '' }));

    await expect(promise).rejects.toThrow('Argument "edit" is required at hook "msg"');
  });
});
