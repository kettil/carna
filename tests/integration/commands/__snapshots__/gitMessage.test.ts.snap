// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command git with git "msg" hook it should work with hooks: child-process 1`] = `
Object {
  "args": Array [
    "run",
    "precarna:git",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command git with git "msg" hook it should work with hooks: child-process 2`] = `
Object {
  "args": Array [
    "--color",
    "--edit",
    "path/to/commit.message",
    "--config",
    "/path/to/project/node_modules/carna/configs/commitlintrc.json",
  ],
  "command": "/path/to/project/node_modules/.bin/commitlint",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command git with git "msg" hook it should work with hooks: child-process 3`] = `
Object {
  "args": Array [
    "run",
    "postcarna:git",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command git with git "msg" hook it should work: child-process 1`] = `
Object {
  "args": Array [
    "--color",
    "--edit",
    "path/to/commit.message",
    "--config",
    "/path/to/project/node_modules/carna/configs/commitlintrc.json",
  ],
  "command": "/path/to/project/node_modules/.bin/commitlint",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;
